File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string func(string s,int l,int r){
        int n = s.length();
        if(r-l == 1){
            if(s[l]!=s[r]){
                return "";
            }
        }
        while(r<n && l>=0 && s[l]==s[r]){
            l--;
            r++;
        }
        // cout<<l+1<<"--"<<r-1<<endl;
        if(l+1 == r-1){
            return s.substr(l+1,1);
        }
        return s.substr(l+1,r-l-1);
    }
    string longestPalindrome(string s) {
        int n = s.length();
        if(n == 1){
            return s;
        }
        string res = "";
        int len = 0;
        for(int i=0;i<n;i++){
            // cout<<"i -- "<<i<<endl;
            string odd = func(s,i,i);
            // cout<<"odd = "<<odd<<endl;
            if(len < odd.length()){
                len = odd.length();
                res = odd;
            }
            string even ;
            if(i<n-1){
                even = func(s,i,i+1);
                if(len < even.length()){
                    len = even.length();
                    res = even;
                }
            }
            // cout<<"even = "<<even<<endl;
            
        }
        return res;
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def oddPal(self,idx:int, s: str, n:int) -> str:
        ans = ""
        l =r= idx

        while l>=0 and r<n:
            if s[l] != s[r]:
                break
            l-=1
            r+=1
        
        return s[l+1:r]

    def evenPal(self, l:int, r:int, s: str, n: int):
        while l>=0 and r<n:
            if s[l] != s[r]:
                break
            l-=1
            r+=1
        
        return s[l+1:r]


    def longestPalindrome(self, s: str) -> str:
        ans = ""
        maxLen = 0
        n = len(s)
        for i in range(n):
            odd = even = ""

            odd = self.oddPal(i,s,n)
            if i<n-1:
                even = self.evenPal(i,i+1,s,n)

            if len(odd) > maxLen:
                ans = odd
                maxLen = len(odd)
            
            if len(even) > maxLen:
                ans = even
                maxLen = len(even)

        return ans